Update Storage samples

tg-msft opened the pull request:
This gets the Hello World samples and structure in place and updates the READMEs.

More samples will be coming.

tg-msft committed:
Update Storage samples

tg-msft committed:
Updating doc warden exclusions for root Storage README

kfarmer-msft commented on sdk/storage/Azure.Storage.Blobs/README.md:
```diff
@@ -33,77 +44,131 @@ Blob storage is designed for:
+BlobContainerClient container = new BlobContainerClient(connectionString, Randomize("sample-container"));
+try
+{
+    container.Create();
```
Given the confusion that has existed over this exact implementation for CreateIfNotExists, I wonder if it should be avoided, and instead use a different operation as an example, like download.

Alternatively, perhaps add a 1-line comment to the effect of avoiding race conditions that exist using a separate existance check.

tg-msft replied:
Since we don't have a `CreateIfNotExists`, I think this is going to be very common for users so I'd like to have it in the README.  I'll add a comment about avoiding race conditions with a separate check.

kfarmer-msft commented on sdk/storage/Azure.Storage.Blobs/README.md:
```diff
@@ -33,77 +44,131 @@ Blob storage is designed for:
-or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any
+This project has adopted the [Microsoft Open Source Code of Conduct][coc].
+For more information see the [Code of Conduct FAQ][coc_faq]
+or contact [opencode@microsoft.com][coc_contact] with any
```
I *think* md supports defining text along with URL for named URLs.  Might reduce some of the bookkeeping here.

tg-msft replied:
It does, but the initial template pulled them all out to the bottom and the team has been sticking with that so I made it consistent.

kfarmer-msft reviewed (commented):
Might suggest a different choice of sample in one case, but otherwise no issues with it.

tg-msft committed:
Sample updates

- Add a public blob so there's something ready to download
- Add CR feedback
- Fix a couple of slow playback issues from a recent change

tg-msft committed:
Add auth samples and fix AD links

maririos commented on sdk/storage/Azure.Storage.Blobs/Changelog.txt:
```diff
@@ -0,0 +1,11 @@
+Release History
```
Should this and `BreakingChanges` be .md? does it matter?

tg-msft replied:
The Storage team has been using these as .txt forever.  I don't think it matters much and we can switch to .md in the future, but we can't change it right now because it's referenced in their NuGet package description.

maririos commented on sdk/storage/Azure.Storage.Blobs/README.md:
````diff
@@ -7,24 +7,35 @@ storage is optimized for storing massive amounts of unstructured data.
 
 ```Powershell
-Install-Package Azure.Storage.Blobs
+dotnet add package Azure.Storage.Blobs --version 12.0.0-preview.1
````
you could do `-IncludePrerelease` to avoid changing the version every time there is a new release

tg-msft replied:
Unfortunately not with `dotnet add package` and floating prerelease versions: https://github.com/NuGet/Home/issues/4699

maririos replied:
:| me not reading...

maririos commented on sdk/storage/Azure.Storage.Blobs/README.md:
```diff
@@ -33,77 +44,131 @@ Blob storage is designed for:
+
+Get started with our [Blob samples][samples]:
+
+1. [Hello World](samples/Sample01a_HelloWorld.cs): Upload, download, and list blobs (or [asynchronously](samples/Sample01b_HelloWorldAsync.cs))
```
Hmm it might be too late to tell people that we also have async APIs in here... Consider adding an explanation in the Examples section where you say what type of examples they can find and that the package also includes async APIs. Something like: https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/keyvault/Azure.Security.KeyVault.Keys#examples

tg-msft replied:
That's a good idea.  Fixed.

maririos commented on sdk/storage/Azure.Storage.Blobs/README.md:
````diff
@@ -33,77 +44,131 @@ Blob storage is designed for:
+
+```c#
+using Azure.Identity;
+using Azure.Security.KeyVault.Secrets;
````
copy past error?

tg-msft replied:
Yes - good catch!  😄

maririos commented on sdk/storage/Azure.Storage.Blobs/README.md:
```diff
@@ -33,77 +44,131 @@ Blob storage is designed for:
+using Azure.Security.KeyVault.Secrets;
+
+// Create a BlobServiceClient that will authenticate through Active Directory
+Uri accountUri = new Uri("https://MYSTORAGEACCOUNT.blob.core.windows.net/");
```
Might be worth adding a comment saying that they will need the Uri of the storage account they previously created/ want to consume...

maririos commented on sdk/storage/Azure.Storage.Blobs/README.md:
````diff
@@ -33,77 +44,131 @@ Blob storage is designed for:
+BlobServiceClient client = new BlobServiceClient(accountUri, new DefaultAzureCredential());
+```
+
+Learn more about enabling Azure Active Directory for authentication with Azure Storage in [our documentation](storage_ad) and [our samples](#next-steps).
````
```suggestion
Learn more about enabling Azure Active Directory for authentication with Azure Storage in [our documentation][storage_ad] and [our samples](#next-steps).
```

maririos replied:
Link is broken. Looks just like formatting

tg-msft replied:
Good catch - fixing in all three.

maririos commented on sdk/storage/Azure.Storage.Blobs/samples/Azure.Storage.Blobs.Samples.Tests.csproj:
```diff
@@ -0,0 +1,19 @@
+  <ItemGroup>
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\src\Azure.Storage.Blobs.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\src\Azure.Storage.Common.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\..\identity\Azure.Identity\src\Azure.Identity.csproj" />
```
Separate form your change, should we reference the package itself? or do something similar like we do with Core that we call .props and according to build config it uses Project or Package?

tg-msft replied:
Maybe?  I don't think this usage matters as much because there's no reference in the shipped packages and we only use Identity for testing that Storage works with AD.  If anyone ever needs to take a dependency on more than TokenCredential we probably should.

jsquire commented on sdk/storage/Azure.Storage.Files/README.md:
````diff
@@ -9,79 +9,178 @@ Windows, Linux, and macOS. Additionally, Azure file shares can be cached on
+[Azure PowerShell][storage_account_create_ps], or the [Azure CLI][storage_account_create_cli].
+Here's an example using the Azure CLI:
+
+```Powershell
````
nit:  This appears to be the Azure CLI, not PowerShell

tg-msft replied:
You can run the Azure CLI from within PS just fine.

jsquire commented on sdk/storage/Azure.Storage.Blobs/README.md:
````diff
@@ -7,24 +7,35 @@ storage is optimized for storing massive amounts of unstructured data.
+[Azure PowerShell][storage_account_create_ps], or the [Azure CLI][storage_account_create_cli].
+Here's an example using the Azure CLI:
+
+```Powershell
````
nit:  This appears to be the Azure CLI, not PowerShell

jsquire commented on eng/.docsettings.yml:
```diff
@@ -63,6 +62,7 @@ known_content_issues:
   - ['sdk/appconfiguration/Azure.ApplicationModel.Configuration/README.md','#5499']
   - ['sdk/appconfiguration/Azure.ApplicationModel.Configuration/tests/Readme.md','#5499']
   - ['sdk/core/Azure.Core/README.md','#5499']
+  - ['sdk/storage/README.md','azure-sdk-tools/issues/42']
```
nit:  Just a double-check....  are there are still content "errors" after your changes?

tg-msft replied:
Yes - this is the top level README which links to the different libraries.  It doesn't have its own Examples, etc.

jsquire commented on sdk/storage/Azure.Storage.Common/samples/Azure.Storage.Common.Samples.Tests.csproj:
```diff
@@ -0,0 +1,21 @@
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Files\src\Azure.Storage.Files.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Queues\src\Azure.Storage.Queues.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\src\Azure.Storage.Common.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\..\identity\Azure.Identity\src\Azure.Identity.csproj" />
```
Should this be a package reference rather than source?

jsquire commented on sdk/storage/Azure.Storage.Files/samples/Azure.Storage.Files.Samples.Tests.csproj:
```diff
@@ -0,0 +1,19 @@
+  <ItemGroup>
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\src\Azure.Storage.Files.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\src\Azure.Storage.Common.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\..\identity\Azure.Identity\src\Azure.Identity.csproj" />
```
Should this be a package reference rather than source?

maririos commented on sdk/storage/Azure.Storage.Common/README.md:
```diff
@@ -10,55 +10,70 @@ and Azure Queues.
 Azure Storage client libraries.
 
-[Source code][source] | [Package (NuGet)][package] | [API reference documentation][rest_docs] | [Product documentation][product_docs]
+[Source code][source] | [Package (NuGet)][package] | [API reference documentation][docs] | [REST API documentation][rest_docs] | [Product documentation][product_docs]
```
Is the idea for the API reference docs to go to generic Azure.Storage instead of Common?

jsquire commented on sdk/storage/Azure.Storage.Queues/samples/Azure.Storage.Queues.Samples.Tests.csproj:
```diff
@@ -0,0 +1,19 @@
+  <ItemGroup>
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\src\Azure.Storage.Queues.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\src\Azure.Storage.Common.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\..\identity\Azure.Identity\src\Azure.Identity.csproj" />
```
Should this be a package reference rather than the raw source?

jsquire reviewed (approved):
I left a couple of thoughts in one or two places that appear elsewhere to avoid repeating things.   If you find value in them, they're likely to apply to the other storage areas as well.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/samples/Azure.Storage.Blobs.Samples.Tests.csproj:
```diff
@@ -0,0 +1,19 @@
+  <ItemGroup>
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\src\Azure.Storage.Blobs.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\src\Azure.Storage.Common.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\..\identity\Azure.Identity\src\Azure.Identity.csproj" />
```
Since it's not referenced by any of the shipping packages, I don't mind keeping it as a source reference.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/samples/Sample01a_HelloWorld.cs:
```diff
@@ -0,0 +1,194 @@
+    /// <summary>
+    /// Basic Azure Blob Storage samples
+    /// </summary>
+    public class Sample01a_HelloWorld : SampleTest
```
I did that at first, but felt it made the samples harder to read.  I was comfortable ignoring them at the cost of a few dozen orphaned 1K text files in the system temp directory.

If I fill my laptop's SSD after frequently running tests for a few weeks, I may revisit this decision.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/samples/Sample01b_HelloWorldAsync.cs:
```diff
@@ -0,0 +1,193 @@
+            }
+
+            // Clean up after the test when we're finished
+            await container.DeleteAsync();
```
Nested try blocks made it harder to read and see the part of the sample I was trying to focus on.

maririos commented on sdk/storage/Azure.Storage.Common/README.md:
```diff
@@ -10,55 +10,70 @@ and Azure Queues.
-Please see the examples for [Blobs][blobs_examples], [Files][blobs_examples],
-or [Queues][blobs_examples] to get started.
+
+Please see the examples for [Blobs][blobs_examples], [Queues][blobs_examples], and [Files][blobs_examples].
```
All these point to the `blob_examples` link

tg-msft replied:
Good catch!  Fixed.

maririos commented on sdk/storage/Azure.Storage.Common/README.md:
```diff
@@ -10,55 +10,70 @@ and Azure Queues.
 ## Next steps
-Get started with [Blobs][blobs_examples], [Files][blobs_examples], or [Queues][blobs_examples].
+
+Get started with our [Common samples](samples) and then continue on with our [Blobs][blobs_samples], [Queues][blobs_samples], and [Files][blobs_samples] samples.
```
Same here :)

tg-msft replied:
Fixed as well

tg-msft commented on sdk/storage/Azure.Storage.Blobs/samples/Sample02_Auth.cs:
```diff
@@ -0,0 +1,211 @@
+    /// For more information, see
+    /// https://docs.microsoft.com/en-us/azure/storage/common/storage-auth
+    /// </summary>
+    public class Sample02_Auth : SampleTest
```
I call it out in the README, but I also thing that's going to be the convention with Storage samples.  We'll group a bunch of related concepts into a single test file.

maririos commented on sdk/storage/Azure.Storage.Files/README.md:
```diff
@@ -93,4 +192,8 @@ additional questions or comments.
 [error_codes]: https://docs.microsoft.com/en-us/rest/api/storageservices/file-service-error-codes
-[samples]: tests/Samples/
\ No newline at end of file
+[samples]: tests/Samples/
```
broken link. 
```suggestion
[samples]: samples/
```

tg-msft committed:
Update sdk/storage/Azure.Storage.Files/README.md

Co-Authored-By: Mariana Rios Flores <mariari@microsoft.com>

maririos commented on sdk/storage/Azure.Storage.Files/README.md:
```diff
@@ -9,79 +9,178 @@ Windows, Linux, and macOS. Additionally, Azure file shares can be cached on
 - Simplify new cloud development projects with shared application settings, diagnostic shares, and Dev/Test/Debug tool file shares.
 
 ## Examples
-### Create a file share
```
In other readme files you were adding the `using` statements too. At least for one example. I found it useful.
i.e. https://github.com/Azure/azure-sdk-for-net/tree/a8650518034730d88beb09956d428a9e2aa08ac9/sdk/storage/Azure.Storage.Blobs#uploading-a-blob

tg-msft replied:
Fixed

tg-msft committed:
Add async examples and PR fixes

maririos commented on sdk/storage/Azure.Storage.Queues/README.md:
```diff
@@ -97,6 +164,13 @@ additional questions or comments.
 [error_codes]: https://docs.microsoft.com/en-us/rest/api/storageservices/queue-service-error-codes
-[samples]: tests/Samples/
\ No newline at end of file
+[samples]: tests/Samples/
```
same here

tg-msft replied:
Fixed

tg-msft committed:
Merge branch 'samples' of https://github.com/tg-msft/azure-sdk-for-net into samples

maririos commented on sdk/storage/README.md:
```diff
@@ -0,0 +1,38 @@
+# Azure Storage libraries for .NET
```
client libraries?

tg-msft replied:
It includes management libraries as well so I left it just "libraries"

maririos replied:
:( makes sense

maririos commented on sdk/storage/Azure.Storage.Common/README.md:
```diff
@@ -10,55 +10,70 @@ and Azure Queues.
 The Azure.Storage.Common library provides infrastructure shared by the other
 Azure Storage client libraries.
 
-[Source code][source] | [Package (NuGet)][package] | [API reference documentation][rest_docs] | [Product documentation][product_docs]
```
NIT: Should the title of this file have the `c` of `common` upper case? :|

tg-msft replied:
Fixed

tg-msft committed:
Small PR feedback on links and casing
